cpu/atmega_common/periph/gpio.c

```diff
@@ -416,6 +416,25 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
     return 0;
 }
 
+int gpio_set_cb(gpio_t pin, gpio_cb_t cb, void *arg)
+{
+    int8_t int_num = _int_num(pin);
+
+    if (int_num < 0) {
+        return -1;
+    }
+
+    if (cb) {
+        config[int_num].cb = cb;
+    }
+
+    if (arg) {
+        config[int_num].arg = arg;
+    }
```

Comment:
@benpicco: I'm referring to this. Let's say I want to set `cb` to `foo`, and `arg` to `NULL`. This is not possible, as only for non-`NULL` values of `arg` it is actually updated.

Maybe you intended to write the following instead:

```C
if (cb) {
    config[int_num].cb = cb;
    config[int_num].arg = arg;
}
```
?

Reply:
> The idea here is that the callback function might be static and in another compilation unit, but one still wants to update the callback argument.

So you want to provide a way to exchange the argument only, without knowing the callback function?

How about this then:

```C
if (cb) {
    config[int_num].cb = cb;
}
config[int_num].arg = arg;
```
?

Reply:
Yes you are right, that is better.
I can't think of a use case where one would want to exchange the isr callback without changing the argument - at least that would seem pretty hacky.

Reply:
please keep race conditions in mind.

```
...
config[int_num].cb = cb;
// GPIO TRIGGERS, new cb gets executed with old arg
config[int_num].arg = arg;
```

Guard with irq_disable/restore, or add a note that this should be done externally?

Reply:
> I can't think of a use case where one would want to exchange the isr callback without changing the argument

How about a driver that gets context info via ```driver_t *ctx = arg;``` in more than one ISR handler, but changes the callback along with going through its state machine? (I can see myself doing that...)

Reply:
> How about a driver that gets context info via driver_t *ctx = arg; in more than one ISR handler, but changes the callback along with going through its state machine? (I can see myself doing that...)

But that driver will certainly have access to the context anyway, and can thus just provide that argument again.

There definitely needs to be a way to set the argument to `NULL`. It is hard to image a scenario in which the callback does different things depending on the parameter, with `NULL` being a special case.

Reply:
Hm, there are two ways to fulfill both requirements (allow to set `arg = NULL` and allow to update the callback without updating the arg) and neither of them is very pretty:

a) 
```C
#define GPIO_UPDATE_CB_NOOP    ((void*) -1)
…
gpio_update_cb(pin, new_cb, GPIO_UPDATE_CB_NOOP);
```

b)
```C
int gpio_update_cb(gpio_t pin, gpio_cb_t cb, void *arg)
{
    if (cb) {
        config[pin].cb = cb;
    }
    
    if (arg || cb == NULL) {
        config[pin].arg = arg;
    }
}
```

But I think I would prefer a).

Reply:
c)

```C
int gpio_update_cb(gpio_t pin, gpio_cb_t cb);
int gpio_update_cb_arg(gpio_t pin, void *arg);
```

Reply:
> ```c
> int gpio_update_cb(gpio_t pin, gpio_cb_t cb, void *arg)
> {
>     if (cb) {
>         config[pin].cb = cb;
>     }
>     
>     if (arg || cb == NULL) {
>         config[pin].arg = arg;
>     }
> }
> ```

Not `(arg || cb != NULL) `?

Reply:
No the idea was to set the arg to null only if you call `gpio_update_cb(pin, NULL, NULL);`

But I think the cleanest solution here is just to have two setters, one for each variable.

Reply:
Sure we really need this? I don't see a use case where someone whats to update the callback only ***and at the same time*** does not have access to the argument. But making the API more complex to fulfill this need definitely makes the API more complex to understand and use, which in turn will be a source of bugs.